### PR TITLE
AWS: add client factory for assume role use case

### PR DESCRIPTION
**Pull request by jackye1995**
@danielcweeks  as discussed in #1844 , separated PR for assume role client factory support.

### CODE REVIEW

File: aws/src/main/java/org/apache/iceberg/aws/AssumeRoleAwsClientFactory.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    AssumeRoleRequest request = AssumeRoleRequest.builder()
+        .roleArn(roleArn)
+        .roleSessionName(genSessionName())
+        .durationSeconds(timeout)
+        .externalId(externalId)
+        .build();
+
+    clientBuilder.credentialsProvider(
+        StsAssumeRoleCredentialsProvider.builder()
+            .stsClient(StsClient.builder().httpClient(HTTP_CLIENT_DEFAULT).build())
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AssumeRoleAwsClientFactory.java:85]: I am not able to reproduce the case that a region must be provided for STS client, please let me know if there is a way to reproduce the scenario.

File: aws/src/integration/java/org/apache/iceberg/aws/AssumeRoleAwsClientFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    }
+
+    try {
+      Namespace namespace = Namespace.of("allowed_" + UUID.randomUUID().toString().replace("-", ""));
+      glueCatalog.createNamespace(namespace);
+      glueCatalog.dropNamespace(namespace);
+    } catch (GlueException e) {
+      LOG.error("fail to create or delete Glue database", e);
+      Assert.fail("create and delete namespace should succeed");
+    }
```
Comment (giovannifumarola) [inline aws/src/integration/java/org/apache/iceberg/aws/AssumeRoleAwsClientFactoryTest.java:123]: you should add a finally to retry to drop the namespace.

File: aws/src/integration/java/org/apache/iceberg/aws/AssumeRoleAwsClientFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    s3FileIO.initialize(assumeRoleProperties);
+    InputFile inputFile = s3FileIO.newInputFile("s3://" + AwsIntegTestUtil.testBucketName() + "/denied/file");
+    try {
+      inputFile.exists();
+      Assert.fail("Access to s3 should be denied");
+    } catch (S3Exception e) {
+      Assert.assertEquals("Should see 403 error code", 403, e.statusCode());
+    }
+
+    inputFile = s3FileIO.newInputFile("s3://" + AwsIntegTestUtil.testBucketName() + "/allowed/file");
```
Comment (giovannifumarola) [inline aws/src/integration/java/org/apache/iceberg/aws/AssumeRoleAwsClientFactoryTest.java:157]: do you delete this file?

Comment (jackye1995) [reply aws/src/integration/java/org/apache/iceberg/aws/AssumeRoleAwsClientFactoryTest.java:157]: This file does not exist, I call `inputFile.exsits()` just to check it has access permission.

**Comment by rdblue** [comment]
Looks good to me, but I haven't been following much of the config discussion so I'll defer to @danielcweeks.

File: aws/src/main/java/org/apache/iceberg/aws/AssumeRoleAwsClientFactory.java (hunk 1 above)
Comment (danielcweeks) [reply aws/src/main/java/org/apache/iceberg/aws/AssumeRoleAwsClientFactory.java:85]: I wouldn't worry about it then.  We have a rather complicated account setup and I'm not sure I know well enough what is producing the issue internally for us.

**Comment by danielcweeks** [comment]
This looks good to me as well, but I think @giovannifumarola comment might still need to be addressed.  Let me know and I'll push it through.

**Review by danielcweeks — APPROVED** [review]
Thanks @jackye1995!

**Pull request merged**